internal/pkg/composable/controller_test.go

```diff
@@ -131,3 +132,73 @@ func TestController(t *testing.T) {
 	require.True(t, ok)
 	assert.Equal(t, "value2", localMap["key1"])
 }
+
+func TestCancellation(t *testing.T) {
+	cfg, err := config.NewConfigFrom(map[string]interface{}{
+		"providers": map[string]interface{}{
+			"env": map[string]interface{}{
+				"enabled": "false",
+			},
+			"local": map[string]interface{}{
+				"vars": map[string]interface{}{
+					"key1": "value1",
+				},
+			},
+			"local_dynamic": map[string]interface{}{
+				"items": []map[string]interface{}{
+					{
+						"vars": map[string]interface{}{
+							"key1": "value1",
+						},
+						"processors": []map[string]interface{}{
+							{
+								"add_fields": map[string]interface{}{
+									"fields": map[string]interface{}{
+										"add": "value1",
+									},
+									"to": "dynamic",
+								},
+							},
+						},
+					},
+					{
+						"vars": map[string]interface{}{
+							"key1": "value2",
+						},
+						"processors": []map[string]interface{}{
+							{
+								"add_fields": map[string]interface{}{
+									"fields": map[string]interface{}{
+										"add": "value2",
+									},
+									"to": "dynamic",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	log, err := logger.New("", false)
+	require.NoError(t, err)
+
+	// try with variable deadlines
+	timeout := 50 * time.Millisecond
+	for i := 1; i <= 10; i++ {
+		t.Run(fmt.Sprintf("test run %d", i), func(t *testing.T) {
+			c, err := composable.New(log, cfg, false)
+			require.NoError(t, err)
+			ctx, cancelFn := context.WithTimeout(context.Background(), timeout)
+			defer cancelFn()
+			err = c.Run(ctx)
```

Comment:
Do we want to have a test case where timeout is 0 or cancel is called explicitly before run?

Reply:
good idea, thanks